src/query/plan/rewrite/join.hpp

```diff
@@ -0,0 +1,547 @@
+    prev_ops_.push_back(&op);
+    op.input()->Accept(*this);
+    RewriteBranch(&op.subquery_);
+    return false;
+  }
+
+  bool PostVisit(Apply & /*op*/) override {
+    prev_ops_.pop_back();
+    return true;
+  }
+
+  bool PreVisit(LoadCsv &op) override {
+    prev_ops_.push_back(&op);
+    return true;
+  }
+
+  bool PostVisit(LoadCsv & /*op*/) override {
+    prev_ops_.pop_back();
+    return true;
+  }
+
+  std::shared_ptr<LogicalOperator> new_root_;
+
+ private:
+  SymbolTable *symbol_table_;
+  AstStorage *ast_storage_;
+  TDbAccessor *db_;
+  // Collected filters, pending for examination if they can be used for advanced
+  // lookup operations (by index, node ID, ...).
+  Filters filters_;
+  // Expressions which no longer need a plain Filter operator.
+  std::unordered_set<Expression *> filter_exprs_for_removal_;
+  std::vector<LogicalOperator *> prev_ops_;
+  std::unordered_set<Symbol> cartesian_symbols_;
+
+  bool DefaultPreVisit() override { throw utils::NotYetImplemented("optimizing join"); }
+
+  void SetOnParent(const std::shared_ptr<LogicalOperator> &input) {
+    MG_ASSERT(input);
+    if (prev_ops_.empty()) {
+      MG_ASSERT(!new_root_);
+      new_root_ = input;
+      return;
+    }
+    prev_ops_.back()->set_input(input);
+  }
+
+  void RewriteBranch(std::shared_ptr<LogicalOperator> *branch) {
+    JoinRewriter<TDbAccessor> rewriter(symbol_table_, ast_storage_, db_);
+    (*branch)->Accept(rewriter);
+    if (rewriter.new_root_) {
+      *branch = rewriter.new_root_;
+    }
+  }
+
+  std::unique_ptr<HashJoin> GenHashJoin(const Cartesian &cartesian) {
+    const auto &left_op = cartesian.left_op_;
+    const auto &left_symbols = cartesian.left_symbols_;
+    const auto &right_op = cartesian.right_op_;
+    const auto &right_symbols = cartesian.right_symbols_;
+
+    auto modified_symbols = cartesian.ModifiedSymbols(*symbol_table_);
+    modified_symbols.insert(modified_symbols.end(), cartesian.left_symbols_.begin(), cartesian.left_symbols_.end());
+
+    std::unordered_set<Symbol> bound_symbols(modified_symbols.begin(), modified_symbols.end());
+    auto are_bound = [&bound_symbols](const auto &used_symbols) {
+      for (const auto &used_symbol : used_symbols) {
+        if (!utils::Contains(bound_symbols, used_symbol)) {
+          return false;
+        }
+      }
+      return true;
+    };
+
+    for (const auto &filter : filters_) {
+      if (filter.type != FilterInfo::Type::Property) {
+        continue;
+      }
+
+      if (filter.property_filter->is_symbol_in_value_ || !are_bound(filter.used_symbols)) {
+        continue;
+      }
+
+      if (filter.property_filter->type_ != PropertyFilter::Type::EQUAL) {
+        continue;
+      }
+
+      if (filter.property_filter->value_->GetTypeInfo() != PropertyLookup::kType) {
+        continue;
+      }
+      auto rhs_lookup = static_cast<PropertyLookup *>(filter.property_filter->value_);
+
+      auto join_condition = static_cast<EqualOperator *>(filter.expression);
```

Comment:
you can use here `auto *` to empasize pointer

Reply:
Added, thanks!